[4.0] Cassiopeia - update card mod chrome

Scrabble96 opened the pull request:
Pull Request for Issue #30678 (second attempt) .

### Summary of Changes
Updated html > layouts > chromes > card.php  (renamed from 'default', see #30729) to display mod-id as the containing element's (address, div, section, etc) `#id` instead of just in the title so that the id displays whether the title is shown or not.

### Testing Instructions

1. Create a new module if none already exists and in the Advanced tab select Cassiopeia's 'card' chrome from the list in 'Module Style' if not already applied.
2. On the front end, use the browser's inspector to inspect this module with title shown and then again where title is not shown
Result: See section "Actual result BEFORE applying this Pull Request" below.
3. Apply the patch of this PR
4. Repeat steps 1. and 2.
See section "Actual result AFTER applying this Pull Request" below.
### Actual result BEFORE applying this Pull Request
module `#id` only appears if title is shown

### Expected result AFTER applying this Pull Request
1. module `#id` always applied to Cassiopeia modules using Cassiopeia 'card' chrome
2. id can now be used for anchors, etc if 1. above applies.

No visible difference on front end for website visitors

### Documentation Changes Required
Note to highlight change

Scrabble96 committed:
Update cardGrey.php

1. Move mod-id to module container to display mod-id whether header/title shown or not
2. Remove h4 from $headerTag
3. Remove h3 from render code

Scrabble96 committed:
Update default.php

1. Move mod-id to module container so that mod-id is shown whether or not header/title is shown
2. Removed 'h4' from $headerTag

Scrabble96 committed:
Create noTitle.php

This option can be used instead of "style-none" in the index.php to allow modules that generally do not have their header/title shown, e.g. main menu, search, but will add an id to the module - which can be used as anchors - and also include any module class suffix added to the module's setup advanced tab.

richard67 commented on templates/cassiopeia/html/layouts/chromes/cardGrey.php:
```diff
@@ -21,34 +21,22 @@
 
 $moduleTag              = $params->get('module_tag', 'div');
-$moduleAttribs          = [];
+$modulePos 		= $module->position;
```
```suggestion
$modulePos              = $module->position;
```

Scrabble96 replied:
I don't believe it! When I viewed it in Github it was all lined up. Aaaargh.

richard67 commented on templates/cassiopeia/html/layouts/chromes/default.php:
```diff
@@ -21,34 +21,22 @@
+$modulePos              = $module->position;
 $moduleAttribs['class'] = $module->position . ' card ' . htmlspecialchars($params->get('moduleclass_sfx'), ENT_QUOTES, 'UTF-8');
-$headerTag              = htmlspecialchars($params->get('header_tag', 'h4'), ENT_QUOTES, 'UTF-8');
+$modId 			= 'mod-' . $module->id;
```
```suggestion
$modId                  = 'mod-' . $module->id;
```

richard67 commented:
@Scrabble96 You should change your editor's settings so it shows spaces and tabs, so you can see what kind of white space is used.

Scrabble96 committed:
Update templates/cassiopeia/html/layouts/chromes/default.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 commented on templates/cassiopeia/html/layouts/chromes/cardGrey.php:
```diff
@@ -21,34 +21,22 @@
 
 $moduleTag              = $params->get('module_tag', 'div');
-$moduleAttribs          = [];
+$modulePos 		= $module->position;
```
```suggestion
$modulePos              = $module->position;
```
Again there are tabs used instead of spaces. Maybe my mistake before. Correcting to space.

Scrabble96 committed:
Update templates/cassiopeia/html/layouts/chromes/cardGrey.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

SharkyKZ commented on templates/cassiopeia/html/layouts/chromes/cardGrey.php:
```diff
@@ -21,34 +21,22 @@
-	<div class="card-body">
-		<?php if ($module->showtitle && $headerClass === 'card-title') : ?>
+$header = '<div class="card-header'. $headerClass .'"><' . $headerTag . '>' . $module->title . '</ . $headerTag .></div>';
+if ($module->content) : ?>
```
Same comment as in previous PR.

Scrabble96 replied:
I made the changes to cardGrey and default but don't know why the cardGrey version has an X next to it.

SharkyKZ replied:
I'm referring to `$module->content` check. There is already a correct check on line 18.

Bakual commented:
Honestly, I don't think we need a new chrome "noTitle". Hiding a title is possible with each module chrome already. There is an option for just that.

Scrabble96 committed:
Update cardGrey.php

Added correct $headerTag element

Scrabble96 committed:
Update default.php

Added $headerTag

Scrabble96 commented:
> 
> 
> Honestly, I don't think we need a new chrome "noTitle". Hiding a title is possible with each module chrome already. There is an option for just that.

In that case, I don't see the reason for having any of the modules as `"style=none"` because:
a. those modules won't have an id added.
b. you can't add a module class suffix (or if you do, it's ignored). This could be confusing (as it was for me, initially) for someone who does not have the user access level to edit the source code but wants to add a known css class to a module.

Perhaps all the modules currently given `"style=none"` in the index.php should be `"style=default"` which would avoid the need for the new 'noTitle' chrome.

Bakual commented:
> In that case, I don't see the reason for having any of the modules as "style=none" because:
a. those modules won't have an id added.
b. you can't add a module class suffix (or if you do, it's ignored). This could be confusing (as it was for me, initially) for someone who does not have the user access level to edit the source code but wants to add a known css class to a module.

That's exactly the purpose of "none". To have only the module output without any chrome around it.
If you need the id or suffix, you can easily change the module style (= chrome) to another one of your choice in the module options.
The chrome specified in the index.php is just the default chrome to be used. It can be overriden in the module options.
And yes, "default" is a bad and isleading choice as a chrome name.

Scrabble96 commented:
> That's exactly the purpose of "none". To have only the module output without any chrome around it.
> If you need the id or suffix, you can easily change the module style (= chrome) to another one of your choice in the module options.
Yes, of course. I haven't used that way as I've always other methods, so I forgot. Sorry.
> The chrome specified in the index.php is just the default chrome to be used. It can be overriden in the module options.
> And yes, "default" is a bad and misleading choice as a chrome name.

Perhaps 'default' should be renamed as 'unstyled'.

Bakual commented:
> Perhaps 'default' should be renamed as 'unstyled'.

Nah, unstyled is wrong as well, as there certainly are styles applied. Actually "default" applies a "card" styling. And I don't even see the difference to "cardGrey".
But I leave that to others who have more experience with design.

Scrabble96 committed:
Update cardGrey.php

Scrabble96 committed:
Update default.php

Removed second content check

Scrabble96 commented:
> 
> Nah, unstyled is wrong as well, as there certainly are styles applied. Actually "default" applies a "card" styling. And I don't even see the difference to "cardGrey".

"cardGrey" has a light grey background in the class "card-header". "card-header" is assigned to the default.php as well, but doesn't  seem to be applied. I can't work out why. I can't see any other differences.

> But I leave that to others who have more experience with design.

Scrabble96 commented:
Ok. What's wrong with my code this time?

richard67 commented:
> Ok. What's wrong with my code this time?

@Scrabble96 You can check that yourself. At the bottom of the PR where the tests are shown, use the "Details" link at the right side or a failed test. In this case drone failed. The "Details" link brings you to a page where on the left hand side is a list of test steps, failed ones in red. In your case now it is "PHPCS", which means PHP code style. Click this and you get on the right hand side the details on this page: [https://ci.joomla.org/joomla/joomla-cms/35590/1/8](https://ci.joomla.org/joomla/joomla-cms/35590/1/8). You will see that the problem is that you use spaces instead of tabs to indent certain lines.

So please follow my advise from above, set up your editor so you can see tabs and spaces and so can distinguish them. And in addition switch off any automatic indent or formatting in your editor (or IDE) settings, or adjust them in a way so that they meet the code style standards documented e.g. here for PHP: [https://developer.joomla.org/coding-standards/php-code.html](https://developer.joomla.org/coding-standards/php-code.html).

Scrabble96 committed:
Update cardGrey.php

Added aria-labelledby back in for module with title and corrected and aria-labelledby to aria-label for no-title modules

Scrabble96 commented:
> 
> So please follow my advise from above, set up your editor so you can see tabs and spaces and so can distinguish them. And in addition switch off any automatic indent or formatting in your editor (or IDE) settings, or adjust them in a way so that they meet the code style standards documented e.g. here for PHP: https://developer.joomla.org/coding-standards/php-code.html.

Can you please clarify whether I have to use tabs instead of spaces?
If I use four spaces instead of a tab, for instance, is that ok or should I use tabs and specify the number of spaces that the tab uses, in which case, is it 4? On looking at the original code, it looks like it. Thank you.

richard67 commented:
> Can you please clarify whether I have to use tabs instead of spaces?
> If I use four spaces instead of a tab, for instance, is that ok or should I use tabs and specify the number of spaces that the tab uses, in which case, is it 4? On looking at the original code, it looks like it. Thank you.

@Scrabble96 If the code style says you have to use tabs then you have to use tabs. 1 tab means one level of indent. 4 spaces are not the same as 1 tab, it only can be made that it looks the same by editor settings. The drone PHPCS log for which I have pasted the link [https://ci.joomla.org/joomla/joomla-cms/35590/1/8](https://ci.joomla.org/joomla/joomla-cms/35590/1/8) and of which I have explained you how to find it clearly tells the line numbers where tabs have to be used.

Scrabble96 committed:
Update default.php

Added aria-labelledby back in for module with title and corrected and aria-labelledby to aria-label for no-title modules
Tabs instead of spaces

Scrabble96 committed:
Update cardGrey.php

Added aria-labelledby back in for module with title and corrected and aria-labelledby to aria-label for no-title modules
Corrected spaces to tabs

richard67 commented:
PHPCS is fine now. Thanks @Scrabble96 .

Scrabble96 commented:
> 
> 
> PHPCS is fine now. Thanks @Scrabble96 .

Thank you. But I'm confused as to why cardGrey has passed checks but default has not. The ONLY difference in the code is line 25 - the class 'card-grey' does not appear in default.php

richard67 commented:
> Thank you. But I'm confused as to why cardGrey has passed checks but default has not. The ONLY difference in the code is line 25 - the class 'card-grey' does not appear in default.php

@Scrabble96 The check is not only done for that one file which is in your commit. It is always done for the complete installation. I.e. after you had checked in the corrected default.php, the test has still failed because cardGrey.php still was not correct. After you had checked that in with the next commit, the test succeeded because now all files were correct. If you had checked in both changed files with one single commit, the test would have been immediately ok with that one commit.

I hope this explains it now.

Scrabble96 committed:
Delete noTitle.php

After consideration and discussion, I agree that this file is not required.

Scrabble96 commented:
> 
> I hope this explains it now.

More or less. Hopefully, if I do this again (if my nerves will stand it) I'll do a better job.

richard67 commented:
@Scrabble96 The more often you do it, the better it will be 😄

Scrabble96 commented:
What happens now?

richard67 commented:
Now volunteers need to test your PR. If it has 2 good tests, it will get status "RTC", which means "ready to commit". Then finally a maintainer will have a look on it and if ok, merge it, and then it will be in the next 4.0 Beta or RC release after that merge.

The 2 tests can't include the author of a PR because it is assumed that a PR's author has already tested his or her code.

Scrabble96 commented:
> 
> 
> Now volunteers need to test your PR. If it has 2 good tests, it will get status "RTC", which means "ready to commit". Then finally a maintainer will have a look on it and if ok, merge it, and then it will be in the next 4.0 Beta or RC release after that merge.
> 
> The 2 tests can't include the author of a PR because it is assumed that a PR's author has already tested his or her code.

Thanks again, @richard67

richard67 commented:
@Scrabble96 Could you update the description of this PR, section "Summary of Changes"? It still mentions the "noTitle.php" chrome, which was meanwhile deleted, and maybe other things have to be adapted to recent changes, too, in that section.

Beside this, I don't fully understand your testing instructions:

> Copy the code and create new chromes (e.g. cardGreyCopy) in J4 Cassiopeia template.

Why does it need to create a copy? Isn't it sufficient just to test the existing chromes modified by this PR?

Scrabble96 commented:
> @Scrabble96 Could you update the description of this PR, section "Summary of Changes"? It still mentions the "noTitle.php" chrome, which was meanwhile deleted, and maybe other things have to be adapted to recent changes, too, in that section.

Done
 
> Beside this, I don't fully understand your testing instructions:
>
> > Copy the code and create new chromes (e.g. cardGreyCopy) in J4 Cassiopeia template.
> 
> Why does it need to create a copy? Isn't it sufficient just to test the existing chromes modified by this PR?

I always use a copy for testing so that I can view that and the original side by side in separate tabs.

richard67 commented:
@Scrabble96 Now after #30729 has been merged, your PR shows conflicts for file `templates/cassiopeia/html/layouts/chromes/default.php`. I think you can resolve them by renaming the file in your branch from `default.php` to `card.php` and then marking the conflict as recolved in your Git client or IDE. Could you do that? Please let me know if I shall help.

Scrabble96 committed:
Rename default.php to card.php

Scrabble96 commented:
> 
> 
> @Scrabble96 Now after #30729 has been merged, your PR shows conflicts for file `templates/cassiopeia/html/layouts/chromes/default.php`. I think you can resolve them by renaming the file in your branch from `default.php` to `card.php` and then marking the conflict as recolved in your Git client or IDE. Could you do that? Please let me know if I shall help.

Hm. Did that, but something went wrong. Sorry, but I'm away for about an hour now.

richard67 commented:
Maybe you had forgotten to mark the conflict as resolved? I will check if I can solve it here now after the rename.

richard67 committed:
Merge branch '4.0-dev' into 4.0-Cassiopeia-update-mod-chromes-and-add-new

richard67 commented:
Done. After the rename it sill was necessary to resolve the conflict.

richard67 commented:
> I always use a copy for testing so that I can view that and the original side by side in separate tabs.

@Scrabble96 That makes maybe sense when you develop locally before making a PR. But it doesn't make any sense for this PR, because when the PR is applied, the file to be copied is already the changed file. So please correct your testing instructions so that they make sense. Thanks in advance.

Scrabble96 commented:
> 
>  So please correct your testing instructions so that they make sense. Thanks in advance.

Is this ok now?

richard67 commented:
> Is this ok now?

Well, step 1 is simply "Apply the patch of this PR", it is on the tester if he or she does this by copying files, like you wrote, or if he or she is using the patchtester for that, or applying a diff with git, there are several ways.

But when applying the patch as first, you can't test the "Actual result BEFORE applying this Pull Request" part anymore.

Many PR's leave it like this or even leave away step 1 because it is more or less clear that a patch has to be applied to see the "Expected result AFTER applying this Pull Request" result. Sometimes they only mention it is it requires additional steps like e.g. running npm to compile some scss.

But if you want to make it perfect, you move step 1 "Apply the patch of this PR" to after the first test, and after that repeat the test. This is how I do it most times. So it would be:

1. Create a new module if none already exists and in the Advanced tab select the changed chrome to be tested from the list in 'Module Style' if not already applied.
2. On the front end, use the browser's inspector to inspect this module with title shown and then again where title is not shown.
Result: See section "Actual result BEFORE applying this Pull Request" below.
3. Apply the patch of this PR.
4. Repeat steps 1 and 2.
Result: See section "Expected result AFTER applying this Pull Request" below.

In the new step 1 (old step 2) it could be helpful to specify which chromes to be tested. The "the changed chrome to be tested" is not clear to everybody when just reading the instructions.

richard67 committed:
Merge remote-tracking branch 'upstream/4.0-dev' into 4.0-Cassiopeia-update-mod-chromes-and-add-new

richard67 commented:
@Scrabble96 Meanwhile PR #30734 which removed CarGrey has been merged. I've allowed myself to update your PR by the latest changes in the 4.0-dev branch, including that one, and solved the "conflict" by using the deleted file. Sounds strange, but that's how it works in git.

Could you update your testing instructions so they don't mention CardGrey anymore?

Thanks in advance.

Scrabble96 commented:
Updated to remove ref to cardGrey

Formatio-hippocampi commented:
I have tested this item :red_circle: unsuccessfully on c614e23e6afa64010b981924054c4bbf1a897bfa<br /><br />Visible differences on frontend:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

Scrabble96 commented:
> 
> 
> I have tested this item 🔴 unsuccessfully on [c614e23](https://github.com/joomla/joomla-cms/commit/c614e23e6afa64010b981924054c4bbf1a897bfa)
> 
Which device and browser are you using?

Formatio-hippocampi commented:
iMac, Firefox 81.

1252spike commented:
I have tested this item :red_circle: unsuccessfully on c614e23e6afa64010b981924054c4bbf1a897bfa<br /><br />Using chrome on windows I see exactly the same error with < class="card-header "> appearing on all the modules before and after the module being tested.



<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

1252spike commented:
![screen shot 2020-09-25 at 13 08 24](https://issues.joomla.org/uploads/1/cfc86bee351534b0f074ce5d83714ecb.jpg)

This is the code I see in inspect<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

SharkyKZ commented on templates/cassiopeia/html/layouts/chromes/card.php:
```diff
@@ -21,34 +21,19 @@
 $moduleAttribs['class'] = $module->position . ' card ' . htmlspecialchars($params->get('moduleclass_sfx'), ENT_QUOTES, 'UTF-8');
-$headerTag              = htmlspecialchars($params->get('header_tag', 'h4'), ENT_QUOTES, 'UTF-8');
+$modId                  = 'mod-' . $module->id;
+$headerTag              = htmlspecialchars($params->get('header_tag', ''), ENT_QUOTES, 'UTF-8');
```
To solve the issues above.

```suggestion
$headerTag              = htmlspecialchars($params->get('header_tag', 'h3'), ENT_QUOTES, 'UTF-8');
```

Scrabble96 replied:
> 
> 
> To solve the issues above.

I don't understand what the 'h3' in line 27 is for. The header tag is chosen in the individual module setup by selecting the header tag in the Advanced tab.

However, I don't see any of the rogue code in Firefox inspector (Windows 10). I have just updated to last night's nightly in case some code had changed, but it is the same.

This is a view on the front end (I have selected the h6 header tag)

![1-card-with-and-without-header](https://user-images.githubusercontent.com/36987494/94278226-7b503580-ff42-11ea-9367-2c9b52154077.jpg)


And this is the inspected code as expected:

![2-card-html](https://user-images.githubusercontent.com/36987494/94278248-81dead00-ff42-11ea-9104-8970ebbe69fd.jpg)

SharkyKZ replied:
This is for getting a default value when no value is set. It's not an issue in your case because you have already saved the module. But it is an issue with modules installed by default which don't have the parameter stored in the database.

Scrabble96 replied:
Ah. OK.

Scrabble96 committed:
Update card.php

Added h3 back into line 27

1252spike commented:
I have not tested this item.<br /><br />Tested on all 'Module Tags' in the advanced tab. Module #id appears in every tag with and without title. Previous error reported no longer exists.

If Module Style -> none then no #id appears. 

Side note: Module Tag also doesn't appear if Module Style -> none option is selected .. is this a desired result?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

1252spike commented:
I have tested this item :white_check_mark: successfully on 2bb9e0c8892f3498f105b3077cde62ca2efa8bcb<br /><br />Tested on all 'Module Tags' in the advanced tab. Module #id appears in every tag with and without title. Previous error reported no longer exists.

If Module Style -> none then no #id appears. 

Side note: Module Tag also doesn't appear if Module Style -> none option is selected .. is this a desired result?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

1252spike commented:
Tested on all 'Module Tags' in the advanced tab. Module #id appears in every tag with and without title. Previous error reported no longer exists.

If Module Style -> none then no #id appears. 

Side note: Module Tag also doesn't appear if Module Style -> none option is selected .. is this a desired result?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

Scrabble96 commented:
> 
> 
> Tested on all 'Module Tags' in the advanced tab. Module #id appears in every tag with and without title. Previous error reported no longer exists.
> 
> If Module Style -> none then no #id appears.
> 
> Side note: Module Tag also doesn't appear if Module Style -> none option is selected .. is this a desired result?
> 
It happens because 'none' means that no chrome is used. On my own custom template (i.e. not Cassiopeia) I don't use `"style=none"` for any modules for exactly that reason.

SharkyKZ commented on templates/cassiopeia/html/layouts/chromes/card.php:
```diff
@@ -21,34 +21,19 @@
 
 $moduleTag              = $params->get('module_tag', 'div');
-$moduleAttribs          = [];
+$modulePos              = $module->position;
```
Unused variable

```suggestion
```

infograf768 replied:
This has not been modified in PR. Changing conversation to Unresolved.

Scrabble96 replied:
I am confused (not difficult) but that line does not appear in the code in my branch.
Should I put it back in without committing it to this PR and then delete it again and commit that?
Or just click 'Commit suggestion' again and hope it works?

Update: I clicked 'Commit suggestion' again.

Scrabble96 committed:
Update templates/cassiopeia/html/layouts/chromes/card.php

Co-authored-by: SharkyKZ <sharkykz@gmail.com>

Formatio-hippocampi commented:
I have tested this item :red_circle: unsuccessfully on 6bdab92f6b88aa5bb972aa2f3fa015b5bcf034d8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

richard67 commented:
@Formatio-hippocampi Thank you for testing. Could you report back what did not work for you so your test was not successful? You should always provide some information when a test failed so we can knwow hat has failed and try to reproduce it. Thanks in advance.

Formatio-hippocampi commented:
was working on update the comment.

richard67 commented:
Thanks. I will be patient 😄

infograf768 commented:
If I understand correctly, this PR only deals with the card chrome, not the other possible chromes layouts.

richard67 commented:
> If I understand correctly, this PR only deals with the card chrome, not the other possible chromes layouts.

That's my understanding, too.

@Formatio-hippocampi Could you check and maybe correct your testing result? It seems your test was successful because for "card" it worked, and the other chromes are not touched by this PR. The "card" was before a recently merged other PR the "default" chrome, testing instructions still refer to that.

Formatio-hippocampi commented:
![Screen Shot 2020-09-26 at 10 30 48](https://user-images.githubusercontent.com/70466955/94336411-6c6d8f80-ffe3-11ea-9465-580230d85a47.png)

Scrabble96 commented:
> 
> 
> > If I understand correctly, this PR only deals with the card chrome, not the other possible chromes layouts.
>  The "card" was before a recently merged other PR the "default" chrome, testing instructions still refer to that.

I have just updated the testing instructions changing 'default' to 'card'

Scrabble96 commented:
> 
> 
> If I understand correctly, this PR only deals with the card chrome, not the other possible chromes layouts.

Yes. There is currently only one mod chrome in the Cassiopeia template which is called 'card'.

Formatio-hippocampi commented:
so:
![94336411-6c6d8f80-ffe3-11ea-9465-580230d85a47](https://user-images.githubusercontent.com/70466955/94336819-409fd900-ffe6-11ea-973f-2404aa3dd635.png)
didn't mean
![Screen Shot 2020-09-26 at 10 50 45](https://user-images.githubusercontent.com/70466955/94336826-4a294100-ffe6-11ea-9d62-496fb3094408.png)
only styles below "cassiopeia"?

Scrabble96 commented:
> 
> 
> so:
> ![94336411-6c6d8f80-ffe3-11ea-9465-580230d85a47](https://user-images.githubusercontent.com/70466955/94336819-409fd900-ffe6-11ea-973f-2404aa3dd635.png)
> didn't mean
> ![Screen Shot 2020-09-26 at 10 50 45](https://user-images.githubusercontent.com/70466955/94336826-4a294100-ffe6-11ea-9d62-496fb3094408.png)
> only styles below "cassiopeia"?

The PR title and the instructions specify 'Cassiopeia'. This change has nothing to do with the system chromes. Perhaps a new PR to apply mod #ids to system chromes would be useful?

richard67 commented:
@Formatio-hippocampi Could you check the pevious comment and if ok for you change your testing result?

I agree it was not 100% clear from reading the testing instructions, but together with the title it makes sense.

Formatio-hippocampi commented:
It made sense to test "from Template", "System" and "Cassiopeia" as the mentioned "none" in test instuctions are below "System".

Its frustrating to waste time cause instructions are written for developers.

Scrabble96 commented:
> 
> 
> It made sense to test "from Template", "System" and "Cassiopeia" as the mentioned "none" in test instuctions are below "System".
> 
> Its frustrating to waste time cause instructions are written for developers.

Sorry. You are right and 'none' is a system chrome. In the Cassiopeia template index.php it uses `"style=card"` and `"style=none"` and I didn't see that 'none' is listed under system. Apologies. I am definitely not a developer; I just wanted to add the mod #id to as many modules in Cassiopeia as possible instead of just the title when displayed.

infograf768 commented:
Cassiopea also uses html5 as a possible chrome in its index.php.
```
			<?php if ($this->countModules('banner')) : ?>
			<div class="grid-child container-banner">
				<jdoc:include type="modules" name="banner" style="html5" />
			</div>
```

Scrabble96 commented:
> 
> 
> Cassiopea also uses html5 as a possible chrome in its index.php.
> 
> ```
> 			<?php if ($this->countModules('banner')) : ?>
> 			<div class="grid-child container-banner">
> 				<jdoc:include type="modules" name="banner" style="html5" />
> 			</div>
> ```

I have now updated the testing instructions to remove reference to 'none'.
To clarify - this PR applies ONLY to the Cassiopeia 'card' chrome

**Expected result AFTER applying this Pull Request**

    module #id always applied to Cassiopeia modules using Cassiopeia 'card' chrome

Quy commented on templates/cassiopeia/html/layouts/chromes/card.php:
```diff
@@ -21,34 +21,18 @@
-	</div>
-</<?php echo $moduleTag; ?>>
+if ($module->showtitle) : ?>
+	<<?php echo $moduleTag; ?> id="<?php echo $modId; ?>" class="<?php echo $moduleAttribs['class'] ?>" aria-labelledby="<?php echo $module->title; ?>">
```
```suggestion
	<<?php echo $moduleTag; ?> id="<?php echo $modId; ?>" class="<?php echo $moduleAttribs['class'] ?>">
```

Scrabble96 replied:
Why remove the aria-labelledby code? I know it's a repeat of the title, but according to [developers.google.com](https://developers.google.com/web/fundamentals/accessibility/semantics-aria/aria-labels-and-relationships)
"aria-labelledby allows us to specify the ID of another element in the DOM as an element's label."

Quy replied:
But it is not pointing to an ID at the moment but has the title value instead.

Scrabble96 committed:
Update templates/cassiopeia/html/layouts/chromes/card.php

Co-authored-by: Quy <quy@fluxbb.org>

Scrabble96 committed:
Update card.php

Revised mod chrome with title showing to have id assigned to header tag (as original card.php rendering) and outer container to be aria-labelledby mod id

gostn commented:
I have tested this item :white_check_mark: successfully on 64bde148295e1140e65f371a9881768cc105ae6d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

Scrabble96 commented:
Hello. This has now had one successful 'human' test. Can two others please test? Many thanks.

richard67 commented:
> Hello. This has now had one successful 'human' test. Can two others please test? Many thanks.

One other is sufficient. It needs 2 tests in total.

AndyGaskell commented:
I have not tested this item.<br /><br />This looks good.  The testing notes were really clear.

Before the patch, the HTML of my module looked like...
```
<div class="sidebar-right card " aria-label="30716 test module">
   <div class="card-body">
        <div class="mod-custom custom" >
            <p>This is a test module for [4.0] Cassiopeia - update card mod chrome #30716</p>
        </div>
    </div>
</div>
```

After the pathc, the HTML of my module looked like...
```
<div id="mod-122" class="sidebar-right card " aria-label="30716 test module">
    <div class="card-body">
        <div class="mod-custom custom" >
            <p>This is a test module for [4.0] Cassiopeia - update card mod chrome #30716</p>
        </div>
    </div>
</div>
```

It's great to have the ID on every module, really useful for template authoring and content styling.

Thanks @Scrabble96 awesome work :) <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

richard67 commented:
@Scrabble96 Just as explanation: As the PR has 2 good tests now, it got the status "RTC", which means "ready to commit". This means sooner or later finally a maintainer will have a look on it, and it's ok it will be merged and the come into the next 4.0 Beta release after that merge.

Scrabble96 commented:
> I have tested this item ✅ successfully on [64bde14](https://github.com/joomla/joomla-cms/commit/64bde148295e1140e65f371a9881768cc105ae6d)
> 
> This looks good. The testing notes were really clear.

Thanks. They took a few revisions and I’ve learnt a lot so it was worth it.

> It's great to have the ID on every module, really useful for template authoring and content styling.
> 
> Thanks @Scrabble96 awesome work :)
> 
You’re very welcome @AndyGaskell :-))

Scrabble96 commented:
> @Scrabble96 Just as explanation: As the PR has 2 good tests now, it got the status "RTC", which means "ready to commit". This means sooner or later finally a maintainer will have a look on it, and it's ok it will be merged and the come into the next 4.0 Beta release after that merge.

Many thanks, @richard67, for all your help and your patience.

HLeithner commented:
Thanks for your pr but I have some problems with it.

1. A module maybe display more then once in this case you have duplicated ids
2. You use the id one time for the aria label and on the other time (when no title exists) as container id, that's confusing

1. I think that can't be solved that might be the reason why we don't have a module id (ymmv)
2. Use an explicit id for the aria label something like `$module->module . '-' . $module->id . '-aria-header'`

But since 1. can't be resolved I'm not sure if it's a good idea to break the html standard

Scrabble96 commented:
> Thanks for your pr but I have some problems with it.
> 
>     1. A module maybe display more then once in this case you have duplicated ids

How often do you use the same module on the same page? If the same module is used on different pages, then the duplicate ID doesn't matter. But if a module is used twice on the same page then there could be a duplicate ID but this theory then also applies to an article with a manually added ID and used more than once on the same page.
 
>     2. You use the id one time for the aria label and on the other time (when no title exists) as container id, that's confusing
> 
I did originally have the id on the outer container in both instances but this was changed to the current option based on input from others and relates to the use of aria-label and aria-labelledby.

> Annotations
> 
>     1. I think that can't be solved that might be the reason why we don't have a module id (ymmv)
> 
>     2. Use an explicit id for the aria label something like `$module->module . '-' . $module->id . '-aria-header'`

The aria-label can't - as far as I know - be used for URL anchors, which was one of the main reasons I raised this PR.

> 
> But since 1. can't be resolved I'm not sure if it's a good idea to break the html standard

I wonder if there's a way of identifying a second/third/... use of a module on a page and adding an additional flag to the id?

richard67 commented:
>     1. A module maybe display more then once in this case you have duplicated ids

@HLeithner Shame on me that I haven't noticed that. Shall I remove RTC?

SharkyKZ commented:
It was the same way before patch.

HLeithner commented:
> It was the same way before patch.

true then it was wrong before too...

HLeithner commented:
> How often do you use the same module on the same page? If the same module is used on different pages, then the duplicate ID doesn't matter. But if a module is used twice on the same page then there could be a duplicate ID but this theory then also applies to an article with a manually added ID and used more than once on the same page.

Doesn't matter how often we should not ship known broken code... but I wouldn't stop it if a11y team says we have to

> 
> I did originally have the id on the outer container in both instances but this was changed to the current option based on input from others and relates to the use of aria-label and aria-labelledby.

I didn't read all comments, is there a reason why? I don't think we should move the id randomly. But anyway if we keep it for a11y it should only be on the header marked as a11y id

> 
> > Annotations
> > ```
> > 1. I think that can't be solved that might be the reason why we don't have a module id (ymmv)
> > 
> > 2. Use an explicit id for the aria label something like `$module->module . '-' . $module->id . '-aria-header'`
> > ```
> 
> The aria-label can't - as far as I know - be used for URL anchors, which was one of the main reasons I raised this PR.

? that was not what I mean. just rename the id to mod_xyz-123-aria-header (or label if it fits better) and only add it to the header tag.

> 
> > But since 1. can't be resolved I'm not sure if it's a good idea to break the html standard
> 
> I wonder if there's a way of identifying a second/third/... use of a module on a page and adding an additional flag to the id?

sadly not because a module can (or should be cached) and the id is part of the complete html cache.

infograf768 commented:
Back to pending<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

Scrabble96 commented:
So, we seem to be back to square one, where the current module chrome only adds an id if there is a module title displayed.
(My PR was designed to add the id whether the title was displayed or not.)
But it would seem that even having an id shown with the title shouldn't be allowed for the reasons @HLeithner pointed out.
>just rename the id to mod_xyz-123-aria-header
>
 will have the same problem with duplicate ids if the module is displayed twice on the page.

HLeithner commented:
> > just rename the id to mod_xyz-123-aria-header
> 
> will have the same problem with duplicate ids if the module is displayed twice on the page.

Yes but at least it's explained why we have a duplicated id... we can add it with javascript (just a joke)

I'm not sure if the "chrome" is cached too maybe someone can look at it, if not then it would be at least possible to use a "random" id per module

AndyGaskell commented:
I was thinking about this, and there seems to be a couple of options...
# Move this from the id property to the class property.  This way having multiples of a module would not be a problem, but you'd still be able to style your module.
# Merge this as is, and create a new issue to deal with the corner case of multiple instances of the same module existing on a page.

HLeithner commented:
> 1. Move this from the id property to the class property.  This way having multiples of a module would not be a problem, but you'd still be able to style your module.

That doesn't work because the aria label needs an id, but for a normal css property a class would be good anyway.

> 2. Merge this as is, and create a new issue to deal with the corner case of multiple instances of the same module existing on a page.

and nobody will every look at it....

AndyGaskell commented:
@HLeithner fair points there.

In that case, perhaps it is a question of which is worse...
1. Modules not having IDs on their div, and modules with titles, if displayed more than once on the page, having the same ID.
2. Modules with titles, if displayed more than once on the page, having the same ID.

brianteeman commented:
As @HLeithner stated a unique ID is required. Its not an option

jiweigert commented:
Which function is responsible for injecting modules Output into the page?
Hooking into that point to create a internal list of already used module-id's on that page,
when a module-id already exist, adding a "-xx" number to that to-beadded module-id?

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

AndyGaskell commented:
@brianteeman Aye, that was understood.

@jiweigert Yea, that sounds promising.

I think there is value in spinning off the whole modules having duplicate IDs is used displayed twice on the page, issue.


This PR only effects modules with titles, so best to discus under that premise.

Current code
Issues: Modules not having IDs on their div, only on titles, if displayed more than once on the page, titles having the same ID.
This looks like...
```
<div class="sidebar-right card " aria-labelledby="mod-122">
    <h3 class="card-header " id="mod-122">30716 test module</h3>
    <div class="card-body">
        <div class="mod-custom custom" >
            <p>This is a test module for [4.0] Cassiopeia - update card mod chrome #30716</p>
        </div>
    </div>
</div>
```

With this PR
Issues: Modules with titles, if displayed more than once on the page, having the same ID.
This looks like...
```
<div id="mod-122" class="sidebar-right card" aria-labelledby="mod-122-title">
    <h3 class="card-header" id="mod-122-title">30716 test module</h3>
    <div class="card-body">
        <div class="mod-custom custom" >
            <p>This is a test module for [4.0] Cassiopeia - update card mod chrome #30716</p>
        </div>
    </div>
</div>
```
...though I did add the title suffix in a couple of places, compared to the PR's code.

Bakual commented:
Just wondering: How can you have the same module instance twice on the page? Afaik you can't. You can have multiple instances of the same module, even with same parameters. But they are still own instances with own IDs.

HLeithner commented:
> Just wondering: How can you have the same module instance twice on the page? Afaik you can't. You can have multiple instances of the same module, even with same parameters. But they are still own instances with own IDs.

just load the same module position multiple times on the page, for example a menu once used for mobile and one for desktop. (only an example)

Bakual commented:
Ah true, didn't think that anyone would have a duplicate module position. Makes sense now.

Llewellynvdm commented:
@HLeithner 
> ... if not then it would be at least possible to use a "random" id per module

A random unpredictable ID defeats the purpose of an id really in terms of targeting it with anything front-end, having a predictable ID seems important. So can't the position and the module ID be used together as the ID? or what am I missing? Certainly you can't have the same module in the same position multiple times.

Scrabble96 commented:
> A random unpredictable ID defeats the purpose of an id really in terms of targeting it with anything front-end, having a predictable ID seems important. So can't the position and the module ID be used together as the ID? or what am I missing? Certainly you can't have the same module in the same position multiple times.
>
@Llewellynvdm That seems an excellent idea to me and a very simple tweak to the code.

richard67 commented:
> Certainly you can't have the same module in the same position multiple times.

@Llewellynvdm That's not true. Theoretically one can show the same menu module for the same menu multiple times on the same position. That could make sense e.g. if the menu has several levels of submenus and in the first menu you show e.g. the top level, the second one the next deeper level and so on. I have no idea if this is practically relevant, but it is possible.

Scrabble96 commented:
Shall I change and submit the new code? If so, perhaps the #id should be on the module container, regardless of whether the title is displayed or not. In which case, how should I best handle the ‘aria-label’ and ‘aria-labelledby’ aspects? Do we need both?

HLeithner commented:
> A random unpredictable ID defeats the purpose of an id really in terms of targeting it with anything front-end, having a predictable ID seems important. So can't the position and the module ID be used together as the ID? or what am I missing? Certainly you can't have the same module in the same position multiple times.

I see no way how to do this. ymmv

Bakual commented:
If you have the same template position multiple times, or manually load a module multiple times, then you get duplicate IDs.
So you just can't set the ID in the chrome. You need to use class.

Scrabble96 commented:
> 
> If you have the same template position multiple times.
>
Why would you do that?

Scrabble96 commented:
I have modified the chrome (locally, only) to include the module position in the id and it works. But this is based on the assumptionn that module positions are only used once. I'll post it here if anyone is interested. I do, however, get this message on Firefox: "Layout was forced before the page was fully loaded. If stylesheets are not yet loaded this may cause a flash of unstyled content." but not on Chrome or Edge.

Bakual commented:
See comment from Harald why a position may be present more than once.
So using the ID attribute is not an option. Even when coupled with the position you still get duplicates.
You can only use the class attribute. There you can insert the module ID as a class which can then be used for CSS styling.

HLeithner commented:
Funny, I got a link to this document today by a customer: https://www.w3.org/TR/WCAG20-TECHS/H93.html IDs have to be unique.

Scrabble96 commented:
>
>There you can insert the module ID as a class which can then be used for CSS styling.
>
But you can't use it for URL anchors which was one of the two reasons I wanted to add an ID to all modules.

HLeithner commented:
Technically it should be possible to inject the file and line position of the <jdoc> into the module attributes and generate a unique id per module and <jdoc> position. That would invalidate our module caching if it's loaded twice on the same page. That can be solved in 2 ways.

1. make a placeholder in the chrome/module layout and do a str_replace on render (we do this on full page cache for example)
2. ignore it and accept that the same module maybe rendered twice (I think that's a less complicated variant and shouldn't hurt to much)

The relevant starting point for a PR is in libraries/src/Document/HtmlDocument.php:849

Bakual commented:
Probably the simplest approach is to just create your own chrome for your own installation where you need the anchor. There you can make sure that module with that chrome is only showing once per page.

For a module chrome shipped with the CMS we just can't be sure that the ID will be unique.

By the way, a PR which tried to achieve a similar thing with a different approach was https://github.com/joomla/joomla-cms/pull/12473

Scrabble96 commented:
> 
> Probably the simplest approach is to just create your own chrome for your own installation where you need the anchor. There you can make sure that module with that chrome is only showing once per page.
> 

For those who would like it, I am happy to share my chrome code which shows both the module position and the module id, but, as @Bakual says, obviously taking into account that any id should be shown only once per page:

`<?php`
`/**`
` * @package     Joomla.Site`
` * @subpackage  Templates.cassiopeia`
` *`
` * @copyright   Copyright (C) 2005 - 2020 Open Source Matters, Inc. All rights reserved.`
` * @license     GNU General Public License version 2 or later; see LICENSE.txt`
` */`

`defined('_JEXEC') or die;`

`use Joomla\Utilities\ArrayHelper;`

`$module  = $displayData['module'];`
`$params  = $displayData['params'];`
`$attribs = $displayData['attribs'];`

`if ($module->content === null || $module->content === '')`
`{`
`	return;`
`}`

`$moduleTag              = $params->get('module_tag', 'div');`
`$modId                  = $module->id;`
`$modPos                 = $module->position;`
`$moduleAttribs          = [];`
`$moduleAttribs['id']    = $modPos . '-' . $modId;`
`$moduleAttribs['class'] = $module->position . ' card ' . htmlspecialchars($params->get('moduleclass_sfx'), ENT_QUOTES, 'UTF-8');`
`$headerTag              = htmlspecialchars($params->get('header_tag', 'h4'), ENT_QUOTES, 'UTF-8');`
`$headerClass            = 'card-header ' . htmlspecialchars($params->get('header_class', ''), ENT_QUOTES, 'UTF-8');`
`$headerAttribs          = [];`
`$headerAttribs['class'] = $headerClass;`

`if ($module->showtitle) :`
`else:`
`	$moduleAttribs['aria-label'] = $module->title;`
`endif;`

`$header = '<' . $headerTag . ' ' . ArrayHelper::toString($headerAttribs) . '>' . $module->title . '</' . $headerTag . '>';`
`?>`
`<<?php echo $moduleTag; ?> <?php echo ArrayHelper::toString($moduleAttribs); ?>>`
`	<?php if ($module->showtitle && $headerClass !== 'card-title') : ?>`
`		<?php echo $header; ?>`
`	<?php endif; ?>`
`	<div class="card-body">`
`		<?php echo $module->content; ?>`
`	</div>`
`</<?php echo $moduleTag; ?>>`

The html looks like this with title:

![pr-30716-with-id-and-module-position-with-title](https://user-images.githubusercontent.com/36987494/95726426-d5eece80-0c70-11eb-8e28-885d4ecdeecb.jpg)


And like this without a title:

![pr-30716-with-id-and-module-position-no-title](https://user-images.githubusercontent.com/36987494/95726152-785a8200-0c70-11eb-8daf-41a229a20403.jpg)


>
> For a module chrome shipped with the CMS we just can't be sure that the ID will be unique.
>

As pointed out earlier, the current 'card' chrome includes a module ID, but only if the title is shown.

> By the way, a PR which tried to achieve a similar thing with a different approach was #12473
>
I hadn't spotted this as it isn't J4.

ChrisHoefliger commented:
Tested successfully on localhost

richard67 commented:
@HLeithner @Bakual @infograf768 What shall we do with this one? It has 2 good tests, but these seems to be some unresolved discussion? Shall I set discussion status in the issue tracker?

gostn commented:
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

gostn commented:
I have tested this item :white_check_mark: successfully on 64bde148295e1140e65f371a9881768cc105ae6d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30716">issues.joomla.org/tracker/joomla-cms/30716</a>.</sub>

HLeithner commented:
https://github.com/joomla/joomla-cms/pull/30716#issuecomment-706283529 Option 1 is the way to go.

gostn commented:
@richard67 you got the answer a week ago

richard67 commented:
@gostn We don't need trolls here.

gostn commented:
> We don't need trolls here

true

Scrabble96 commented:
> 
> 
> I have tested this item ✅ successfully on [64bde14](https://github.com/joomla/joomla-cms/commit/64bde148295e1140e65f371a9881768cc105ae6d)
> This comment was created with the [J!Tracker Application](https://github.com/joomla/jissues) at [issues.joomla.org/tracker/joomla-cms/30716](https://issues.joomla.org/tracker/joomla-cms/30716).

@gostn this PR was effectively put on hold due to the use of IDs in the chrome which would create duplicate IDs if the same module was used twice on the page, which is not allowed in good website page setup. There was a suggestion to add the 'discussion' tag to this PR, but that hasn't happened. I don't have the authority to do that. My original reason for opening this PR was two-fold: that the ID could be used for in-page anchors and that the ID would be shown whether the title was shown or not. Due to the problem with duplicate IDs this PR doesn't seem viable as it stands.

Frustratingly, before HTML5 the name attribute could be used in for anchors in links but it has now been deprecated in favour of the use of IDs, which brings us back to square one. I wonder when they (whoever 'they' are) deprecated its use whether they considered the possibility of duplicate IDs on a page?

gostn commented:
@Scrabble96 thanks for your friendly information.

hans2103 commented:
@Scrabble96 This PR of your seems to have a merge conflict and cannot be merged into Joomla. 
Can you rewrite your PR?

I have a suggestion. 

https://github.com/joomla/joomla-cms/blob/3b82bf129322cf473f10a1856d4ec385895ff1ad/templates/cassiopeia/html/layouts/chromes/card.php#L24-L25

To make sure a module always has an `id`, rewrite the two lines above into the three lines below:
```
$moduleAttribs          = [];
$moduleAttribs['id'] = 'mod-' . $module->id;
$moduleAttribs['class'] = $module->position . ' card ' . htmlspecialchars($params->get('moduleclass_sfx'), ENT_QUOTES, 'UTF-8');
```

Two prevent duplicate ID's we have to adjust a second part of this file too:
https://github.com/joomla/joomla-cms/blob/3b82bf129322cf473f10a1856d4ec385895ff1ad/templates/cassiopeia/html/layouts/chromes/card.php#L36-L45

To prevent duplicate ID's we have to adjust the name of the id for the header into:
```
// Only add aria if the moduleTag is not a div
if ($moduleTag !== 'div')
{
	if ($module->showtitle) :
		$moduleAttribs['aria-labelledby'] = 'mod-title-' . $module->id;
		$headerAttribs['id']              = 'mod-title-' . $module->id;
	else:
		$moduleAttribs['aria-label'] = $module->title;
	endif;
}
```


With these two changes based on the current implementation of template/cassiopeia/html/layouts/chromes/card.php the PR is  good to go I think. 
What do you think?

HLeithner commented:
As already mentioned if the module is cached (and I'm not wrong) the id is still duplicated.

Scrabble96 commented:
> @Scrabble96 This PR of your seems to have a merge conflict and cannot be merged into Joomla.
> Can you rewrite your PR?
> 
@hans2103 I'll have a look in the next few days but, as @HLeithner says, there is a problem with cached modules so I don't think this is going to work.

Scrabble96 commented:
I'm going to close this PR due to the cache conflict.
